### PR TITLE
Proposition add some Python types

### PR DESCRIPTION
**Pull request by darnuria**
Tl;DR it's low priority mostly opened to share knowledge and create a discussion about usage of types in Python
 
Openned this PR to checkout what you think about types in Python, it's not evaluated by CPython but helps IDE and static analysis tools to find out buts. It's very lightweight not like Java and it's not mandatory can be done gradually where it matters.

It unlock usage of tools like [mypy](https://mypy.readthedocs.io/en/stable/) we will use python typing in our own sources for next developments @coopaname and will types old sources gradually where applicable.

Some reading about it: <https://medium.com/analytics-vidhya/type-annotations-in-python-3-8-3b401384403d>

Not totally dependants of types but improve quality of [Pysa](https://engineering.fb.com/2020/08/07/security/pysa/) analysis.

In short the main goal is to find bugs before production.

### CODE REVIEW

File: coopaname_custom/models/event.py [hunk 1]
```diff
@@ -102,7 +103,7 @@ def sync_co_organizer_registrations(self, new_co_organizer_id):
             new_registration.confirm_registration()
 
     @api.model
-    def create(self, vals):
+    def create(self, vals) -> Event:
```
Comment (robinkeunen) [inline coopaname_custom/models/event.py:106]: It actually returns a RecordSet of model `event.event` rather than Event class.

File: coopaname_custom/models/hr_employee.py [hunk 2]
```diff
@@ -33,7 +34,7 @@ class Employee(models.Model):
     )
 
     @api.model
-    def create(self, values):
+    def create(self, values) -> Employee:
```
Comment (robinkeunen) [inline coopaname_custom/models/hr_employee.py:37]: same comment.

**Review by robinkeunen — CHANGES_REQUESTED** [review]
The annotations are not correct. :-)

As for the usage of annotations, I never used them. I'm not philosophically opposed to using them, I like code self-documenting itself. I'm not sure wether mypy can cope with Odoo's code base to analyse types distributed accross modules and repositories.

File: coopaname_custom/models/event.py (hunk 1 above)
Comment (darnuria) [reply coopaname_custom/models/event.py:106]: Oh thanks that's totally right will fix it.

**Comment by darnuria** [comment]
About annotation on large codebase, if someone did type mapping or if odoo upstream them it's quite possible django has stub types  <https://github.com/typeddjango/django-stubs>.

I found somebody who is busy working on that here: <https://github.com/trinhanhngoc/odoo-stubs>

It's almost like TypeScript types stubs.

I think it's a good move to help in adoption of odoo because like you self it's leverage self documenting code.
Also help a lot IDE in code completion even in very dynamic codebase.

**Comment by robinkeunen** [comment]
What I'd suggest, is to start using it in coopaname_custom module. If it proves useful, start using it in odoo-cae after having a discussion with GRAP.

By the way, `coopaname_custom`module is destined to be moved out of this repo in a specific coopaname repo or in [coopiteasy/cie-custom](https://github.com/coopiteasy/cie-custom).

**Comment by robinkeunen** [comment]
Ow the stubs led me here: [pycharm-odoo](https://github.com/trinhanhngoc/pycharm-odoo), I'll definitely take a look!

File: coopaname_custom/models/event.py (hunk 1 above)
Comment (darnuria) [reply coopaname_custom/models/event.py:106]: Looked into odoo source code and doc it look like that `RecordSet` is not explicitly defined as a class it's just an ordered collection of something who inherit `model.BaseModel` if I read correctly the docs.

In Odoo stubs the author chose [this signature](https://github.com/trinhanhngoc/odoo-stubs/blob/171cb8f0c2d9427bfe4e5b8d7a5627afc0546349/odoo-stubs/models.pyi#L199)

Citation from <https://github.com/trinhanhngoc/odoo-stubs/> by trinhanhngoc
```python
    def create(self, vals_list: Any) -> BaseModel: ...
```

And about `event.event` is it defined here: <https://github.com/OCA/event> ?

Didn't find real class behind (searched not too hard will spend some time looking) so maybe leaving out the «interface» here `BaseModel` is better than nothing, except you have some insight will need some check at runtime to get the right type. 

See what you said with types and odoo, not impossible but involve some corners case due to ultra dynamic object gym.

**Comment by darnuria** [comment]
Hmm interresting the discussion and failure in CI lead me to the point that [CI](https://travis-ci.org/github/odoo-cae/odoo-addons-hr-incubator/jobs/743385751#L240) use python 3.5.6 it's somewhat odd it's not listed in [python active releases](https://www.python.org/downloads/).

[python 3.5.6](https://www.python.org/downloads/release/python-356/) and all 3.5 is in end-of-life status at least an little update to 3.5.10 would be better (last release on 3.5).

Look like there is no apparent blocking point to bump minimal version from [odoo install doc](https://www.odoo.com/documentation/12.0/setup/install.html#id12) but as I looked odoo.sh look like it's stuck in `3.5` channel [src](https://www.odoo.com/documentation/user/12.0/odoo_sh/advanced/containers.html) (if doc is OK) so no big bump to 3.7 + expectable at least now. :o

**Comment by darnuria** [comment]
Closing due to inactivity!
